hooks/request/db.go

```diff
@@ -1,44 +1,123 @@
 package request
 
 import (
-	"encoding/hex"
+	"encoding/base64"
+	"fmt"
+
+	"gorm.io/driver/mysql"
+	"gorm.io/driver/postgres"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
 
 	"github.com/bandprotocol/chain/x/oracle/types"
 )
 
-type Request struct {
-	RequestID      types.RequestID      `db:"request_id, primarykey" json:"request_id"`
-	OracleScriptID types.OracleScriptID `db:"oracle_script_id" json:"oracle_script_id"`
-	Calldata       string               `db:"calldata" json:"calldata"`
-	MinCount       uint64               `db:"min_count" json:"min_count"`
-	AskCount       uint64               `db:"ask_count" json:"ask_count"`
-	ResolveTime    int64                `db:"resolve_time" json:"resolve_time"`
+func initDb(driverName, dataSourceName string) *gorm.DB {
+	var db *gorm.DB
+	var err error
+
+	switch driverName {
+	case "sqlite3":
+		db, err = gorm.Open(sqlite.Open(dataSourceName), &gorm.Config{
+			SkipDefaultTransaction: true,
+		})
+		if err != nil {
+			panic(fmt.Errorf("failed to connect to SQLite: %w", err))
+		}
+	case "postgres":
+		db, err = gorm.Open(postgres.Open(dataSourceName), &gorm.Config{
+			SkipDefaultTransaction: true,
+		})
+		if err != nil {
+			panic(fmt.Errorf("failed to connect to PostgreSQL: %w", err))
+		}
+	case "mysql":
+		db, err = gorm.Open(mysql.Open(dataSourceName), &gorm.Config{
+			SkipDefaultTransaction: true,
+		})
+		if err != nil {
+			panic(fmt.Errorf("failed to connect to MySQL: %w", err))
+		}
+
+	default:
+		panic(fmt.Sprintf("unknown driver %s", driverName))
+	}
+	if err = db.AutoMigrate(&Request{}, &RequestedValidator{}, &RawReport{}, &RawRequest{}, &Report{}); err != nil {
+		panic(fmt.Errorf("unable to auto-migrate DB: %w", err))
+	}
+
+	return db
 }
 
-func (h *Hook) insertRequest(requestID types.RequestID, oracleScriptID types.OracleScriptID, calldata []byte, askCount uint64, minCount uint64, resolveTime int64) {
-	err := h.trans.Insert(&Request{
-		RequestID:      requestID,
-		OracleScriptID: oracleScriptID,
-		Calldata:       hex.EncodeToString(calldata),
-		MinCount:       minCount,
-		AskCount:       askCount,
-		ResolveTime:    resolveTime,
-	})
-	if err != nil {
-		panic(err)
+func (h *Hook) insertRequests(requests []types.QueryRequestResponse) {
+	var dbRequests []Request
+	for _, request := range requests {
+		dbRequests = append(dbRequests, GenerateRequestModel(request))
+	}
+
+	h.trans.Create(&dbRequests)
+}
+
+func (h *Hook) insertReports(reportMap map[types.RequestID][]types.Report) {
+	var results []Report
+	for requestID, reports := range reportMap {
+		for _, report := range reports {
+			results = append(results, GenerateReportModel(requestID, report))
+		}
+	}
+
+	h.trans.Model(&Report{}).Create(&results)
```

Comment:
Is there a case that we cannot insert new report in DB? (request of this report has been deleted from DB)

Reply:
I have added a condition to check whether request exists or not.